sgrep/lib/parse_rules.ml

```diff
@@ -52,37 +57,36 @@ let parse file =
                let languages = langs |> List.map (function
                 | `String s ->
                   (match Lang.lang_of_string_opt s with
-                  | None -> error (spf "unsupported language: %s" s)
+                  | None -> raise (InvalidLanguageException (id, (spf "unsupported language: %s" s)))
                   | Some l -> l
                   )
-                | _ -> error (spf "expecting a string for languages")
+                | _ -> raise (InvalidRuleException (id, (spf "expecting a string for languages")))
                ) in
                let lang =
                  match languages with
-                 | [] -> error "we need at least one language"
+                 | [] -> raise (InvalidRuleException (id, "we need at least one language"))
                  | x::_xs -> x
                in
                let pattern =
                  (* todo? call Normalize_ast.normalize here? *)
                  try Parse_generic.parse_pattern lang pattern
                  with exn ->
-                   error (spf "could not parse the pattern %s (exn = %s)"
-                            pattern (Common.exn_to_s exn))
+                   raise (InvalidPatternException (id, pattern, (Lang.string_of_lang lang), (Common.exn_to_s exn)))
                in
-               let severity = severity_of_string sev in
+               let severity = (severity_of_string id sev) in
                { R. id; pattern; message; languages; severity }
              | x ->
                pr2_gen x;
-               error "wrong rule fields"
+               raise (InvalidYamlException "wrong rule fields")
              )
           | x ->
               pr2_gen x;
-              error "wrong rule fields"
+              raise (InvalidYamlException "wrong rule fields")
          )
-      | _ -> error "missing rules entry"
+      | _ -> raise (InvalidYamlException "missing rules entry as top-level key")
       )
   | Result.Error (`Msg s) ->
-    failwith (spf "sgrep_lint: could not parse %s (error = %s)" file s)
+    raise (UnparsableYamlException s)
```

Comment:
note that this yaml file is always generated by sgrep-lint; regular users are not supposed to feed sgrep-core directly with yaml file (except me when I test things), so those errors should be caught by the sgrep-lint wrapper anyway.

Reply:
True